massa-signature/src/signature_impl.rs

```diff
@@ -321,47 +490,76 @@ impl<'de> ::serde::Deserialize<'de> for KeyPair {
     }
 }
 
+#[allow(missing_docs)]
 /// Public key used to check if a message was encoded
 /// by the corresponding `PublicKey`.
 /// Generated from the `KeyPair` using `SignatureEngine`
+#[transition::versioned(versions("0", "1"))]
 #[derive(Clone, Copy, PartialEq, Eq)]
 pub struct PublicKey(ed25519_dalek::PublicKey);
 
 const PUBLIC_PREFIX: char = 'P';
 
+#[transition::impl_version(versions("0", "1"), structures("PublicKey"))]
+impl PublicKey {
+    /// Size of a public key
+    pub const PUBLIC_KEY_SIZE_BYTES: usize = ed25519_dalek::PUBLIC_KEY_LENGTH;
+}
+
 #[allow(clippy::derived_hash_with_manual_eq)]
 impl std::hash::Hash for PublicKey {
     fn hash<H: Hasher>(&self, state: &mut H) {
-        self.0.as_bytes().hash(state);
+        match self {
+            PublicKey::PublicKeyV0(pubkey) => pubkey.hash(state),
+            PublicKey::PublicKeyV1(pubkey) => pubkey.hash(state),
+        }
     }
 }
 
 impl PartialOrd for PublicKey {
     fn partial_cmp(&self, other: &PublicKey) -> Option<Ordering> {
-        self.0.as_bytes().partial_cmp(other.0.as_bytes())
+        match (self, other) {
+            (PublicKey::PublicKeyV0(pubkey), PublicKey::PublicKeyV0(other_pubkey)) => {
+                pubkey.0.as_bytes().partial_cmp(other_pubkey.0.as_bytes())
+            }
+            (PublicKey::PublicKeyV0(pubkey), PublicKey::PublicKeyV1(other_pubkey)) => {
+                pubkey.0.as_bytes().partial_cmp(&other_pubkey.0.as_bytes())
+            }
+            (PublicKey::PublicKeyV1(pubkey), PublicKey::PublicKeyV0(other_pubkey)) => {
```

Comment:
if used in BtreeMap (for instance), it is required to have total ordering which is not the case here...

Reply:
`PublicKey` itself itself isn't, but `NodeId` is and it is nothing but a wrapper around a `PublicKey`:

```rust
/// `NodeId` wraps a public key to uniquely identify a node.
#[derive(
    Clone, Copy, Eq, Hash, Ord, PartialEq, PartialOrd, SerializeDisplay, DeserializeFromStr,
)]
pub struct NodeId(PublicKey);
```

Reply:
For info: what was asked here was to have ordering of versions too

Reply:
There is actually no need to explicitly specify the version ordering, it is done naturally by the initial version byte, you can see the update in https://github.com/massalabs/massa/pull/3876/commits/455691a6135627e7c487a33ce3d666492b552fbd